### PR TITLE
Account for missing keyword in fn/struct definition

### LINKED ISSUE #38911: Confusing error message: "can't qualify macro invocation with `pub`" when a inherent method lacks "fn"
Both stable 1.14 and nightly 2017-01-06.
[playpen](https://is.gd/Zwvabj)
```rust
struct A {}

impl A {
    pub get(self) {}
}
```
(**Edit:** Above example is fixed as of 1.20.0, but [free functions are affected](https://github.com/rust-lang/rust/issues/38911#issuecomment-331317518).)
```text
rustc 1.14.0 (e8a012324 2016-12-16)
error: can't qualify macro invocation with `pub`
 --> <anon>:4:5
  |
4 |     pub get(self) {}
  |     ^^^
  |
  = help: try adjusting the macro to put `pub` inside the invocation

error: expected one of `!` or `::`, found `(`
 --> <anon>:4:12
  |
4 |     pub get(self) {}
  |            ^

error: aborting due to previous error
```
while the actual issue is that the definition of `get` is missing the `fn` keyword.

### PR DESCRIPTION
**Pull request by estebank**
Fix #38911.

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Actually, looking more closely, I think this is overkill. I'll leave a more detailed comment with an alternative shortly.

**Comment by nikomatsakis** [comment]
Currently, after failing to parse, the parser invokes this commonly used method:

https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6267-L6275

This is intended to look for a macro use like `path!` -- but right now it decides it has seen a macro quite early, based [only on whether the next token can start a path](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6276). Presuming it can, we [error if we saw a `pub`](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6280), [parse the path](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6285),  and then [expect a `!`](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6286). **This is where the error will occur in our bad examples.**

I think we should customize what happens when we "expect" this `!` but do not find it:

To start, we can look at the path. If it was multipart, like `a::b`, then I think this could only be a macro-use, perhaps one where the `!` was forgotten, and we could diagnose it that way. (Or, at least, we could do this if the next token is `(`.)

If the path is a single identifier, we can then examine the next character: if it is `(`, we could suggest a function. If it is `{`, we could suggest a struct. Or, we could just issue something generic like "expected an item like a `fn` or a macro use like `foo!`".

**Comment by estebank** [comment]
It is reasonable to go ahead and do that change, but I also feel there's value in modifying the current PR to keep doing what it is doing and returning `Ok` with the fn or struct that could be parsed, instead of the `Err`. That way parsing can continue and present other errors down the line, instead of the current behavior of having a failure parsing a macro and then a failure parsing the opening brace or paren. What do you think?

**Comment by nikomatsakis** [comment]
@estebank that's a good point. On the other hand, I wonder, could we get a similar effect by just skipping over entire token trees?

I feel like only being able to recover when the input is **perfect** is perhaps not ideal.

**Comment by nikomatsakis** [comment]
though man but mismatched braces are the worst sort of parse errors =)

**Comment by nikomatsakis** [comment]
In any case, I could go either way. But I'd like to see us able to recover even from something like this:

```rust
pub foo(x: u32) {
    x->foo(); // whoops, too much C++
}
```

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/41282 already implemented this recovery for trait and impl items, it can be adapted to normal items as well.
Saving/restoring parser state should be avoided if possible.

**Comment by estebank** [comment]
I have modified the code to not perform the rewind. The outwards behavior now is similar to what it already was (the parser will not let the compiler to continue, but it will try to consume the continuing code in order to identify what it was meant (fn or struct), regardless of code errors (just looks for enclosed `(...)` and `{...}` sections).

File: src/test/ui/suggestions/pub-ident-fn-2.stderr [hunk 1]
```diff
@@ -0,0 +1,13 @@
+error: missing `fn` for method definition
+  --> $DIR/pub-ident-fn-2.rs:11:4
+   |
+11 | pub foo(s: usize) { bar() }
+   |    ^
+   |
+help: add `fn` here to parse `foo` as a public method
+   |
+11 | pub fn foo(s: usize) { bar() }
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/pub-ident-fn-2.stderr]: that is pretty slick =)

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -6225,7 +6275,65 @@ impl<'a> Parser<'a> {
             return Ok(Some(macro_def));
         }
 
-        self.parse_macro_use_or_failure(attrs,macros_allowed,attributes_allowed,lo,visibility)
+        // Verify wether we have encountered a struct or method definition where the user forgot to
+        // add the `struct` or `fn` keyword after writing `pub`: `pub S {}`
+        if visibility == Visibility::Public &&
+            self.check_ident() &&
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: Isn't it possible these days to have *paths* to macros? e.g., `foo::bar!`?

cc @jseyfried @petrochenkov

**Review by nikomatsakis — COMMENTED** [review]
This all looks quite cool. I do have one concern about the recovery condition I'd like to get resolved though.

File: src/libsyntax/parse/parser.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: The way the fall back happens, if the code isn't either `pub foo (` or `pub foo {`, the current macro handling will happen. In the case of `pub foo::bar` this new branch will not be executed. As long as `parse_macro_use_or_failure` does the right thing with paths, this should be fine :)

**Comment by estebank** [comment]
@nikomatsakis this PR is ballooning in size with multiple independent changes that are kind or related but not entirely. There's a very small one liner suggestion underline fix, emitting inappropriate `DocComment` location error but allow it for the parser to continue parsing the struct,  making block parsing errors recoverable, other than the original intention of the PR which was to add a suggestion if the parser encounters `pub ident (` or `pub ident {`.

Because of these changes, in the case of [`compile-fail-fulldeps/proc-macro/derive-bad.rs`](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail-fulldeps/proc-macro/derive-bad.rs) (which uses [`compile-fail-fulldeps/proc-macro/auxiliary/derive-bad.rs`](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail-fulldeps/proc-macro/auxiliary/derive-bad.rs)) the parsing errors from the `derive` are emitted and not seen by `libproc_macro` as a failure, so we get the internal parsing errors, that somebody who can't modify the proc macro can't do anything about, and we don't cause the actual expected message. For the second issue I'll try to find a way to emit it, for the first, is it ok to leak these messages?

As for the first paragraph, should I try to split this in a bunch of different dependent PRs?

**Comment by estebank** [comment]
@nikomatsakis let me know if the output changes are reasonable or if I should keep working on this PR to get to minimal changes from the status quo.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+

@estebank seems good to me.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -6236,7 +6279,65 @@ impl<'a> Parser<'a> {
...
+                err.span_suggestion_short(sp, &msg, " struct ".into());
+                return Err(err);
+            } else if self.look_ahead(1, |t| *t == token::OpenDelim(token::Paren)) {
+                let ident = self.parse_ident().unwrap();
+                self.consume_block(token::Paren);
+                let (kw, kw_name, ambiguous) = if self.check(&token::RArrow) ||
+                    self.check(&token::OpenDelim(token::Brace))
+                {
+                    ("fn", "method", false)
+                } else if self.check(&token::Colon) {
```
Comment (Xiretza) [inline src/libsyntax/parse/parser.rs:6311]: @estebank sorry for the enormous necropost, but do you know what the idea behind this was? It matches `pub foo(stuff):` and suggests adding `struct`, but where does that colon come from? Is it possible this was intended to be `token::Semi` instead? The behaviour isn't covered by any tests, unfortunately.

**Pull request merged**